src/types/unsolicited_response.rs

```diff
@@ -0,0 +1,309 @@
+                        Ok(owncode) => Ok(UnsolicitedResponse::Bye {
+                            code: Some(owncode),
+                            information: info,
+                        }),
+                        _ => Err(response),
+                    }
+                } else {
+                    Ok(UnsolicitedResponse::Bye {
+                        code: None,
+                        information: info,
+                    })
+                }
+            }
+            Response::Fetch(id, ref attributes) => {
+                match AttributeValue::try_from_imap_proto_vec(attributes) {
+                    Ok(attrs) => Ok(UnsolicitedResponse::Fetch {
+                        id,
+                        attributes: attrs,
+                    }),
+                    _ => Err(response),
+                }
+            }
+            _ => Err(response),
+        }
+    }
+}
+
+/// Owned version of ResponseCode that wraps a subset of [`imap_proto::ResponseCode`]
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum ResponseCode {
+    /// Highest ModSeq in the mailbox, [RFC4551](https://tools.ietf.org/html/rfc4551#section-3.1.1)
+    HighestModSeq(u64),
+    /// Next UID in the mailbox, [RFC3501](https://tools.ietf.org/html/rfc3501#section-2.3.1.1)
+    UidNext(Uid),
+    /// Mailbox UIDVALIDITY, [RFC3501](https://tools.ietf.org/html/rfc3501#section-2.3.1.1)
+    UidValidity(u32),
+    /// Sequence number of first message without the `\\Seen` flag
+    Unseen(Seq),
+}
+
+impl<'a> TryFrom<&ImapProtoResponseCode<'a>> for ResponseCode {
+    type Error = ParseError;
+
+    fn try_from(val: &ImapProtoResponseCode<'a>) -> Result<Self, Self::Error> {
+        match val {
+            ImapProtoResponseCode::HighestModSeq(seq) => Ok(ResponseCode::HighestModSeq(*seq)),
+            ImapProtoResponseCode::UidNext(uid) => Ok(ResponseCode::UidNext(*uid)),
+            ImapProtoResponseCode::UidValidity(uid) => Ok(ResponseCode::UidValidity(*uid)),
+            ImapProtoResponseCode::Unseen(seq) => Ok(ResponseCode::Unseen(*seq)),
+            unhandled => Err(ParseError::Unexpected(format!("{:?}", unhandled))),
+        }
+    }
+}
+
+/// Owned version of AttributeValue that wraps a subset of [`imap_proto::AttributeValue`].
+#[derive(Debug, Eq, PartialEq)]
+#[non_exhaustive]
+pub enum AttributeValue {
+    /// Message Flags
+    Flags(Vec<Flag<'static>>),
+    /// Message ModSequence, [RFC4551](https://tools.ietf.org/html/rfc4551#section-3.3.2)
+    ModSeq(u64),
+    /// Message UID, [RFC3501](https://tools.ietf.org/html/rfc3501#section-2.3.1.1)
+    Uid(Uid),
+}
+
+impl<'a> TryFrom<&ImapProtoAttributeValue<'a>> for AttributeValue {
+    type Error = ParseError;
+
+    fn try_from(val: &ImapProtoAttributeValue<'a>) -> Result<Self, Self::Error> {
+        match val {
+            ImapProtoAttributeValue::Flags(flags) => {
+                let v = flags.iter().map(|v| Flag::from(v.to_string())).collect();
```

Comment:
Given the number of places we do this iter+map+Flag::from, maybe we should have a helper for it?

Reply:
Yay helpers! I will do this. :)